hathor/transaction/resources/transaction.py

```diff
@@ -67,17 +67,21 @@ def get_tx_extra_data(tx: BaseTransaction) -> Dict[str, Any]:
 
             # We need to get the token_data from the current tx, and not the tx being spent
             token_uid = tx2.get_token_uid(tx2_out.get_token_index())
-            for out in tx.outputs:
-                out_token_uid = tx.get_token_uid(out.get_token_index())
-                if out_token_uid == token_uid:
-                    output['decoded']['token_data'] = out.token_data
-                    break
+            if token_uid == settings.HATHOR_TOKEN_UID:
+                # If it's an HTR output, the token_data is 0
+                output['decoded']['token_data'] = 0
             else:
-                # This is the case when the token from the input does not appear in the outputs
-                # This case can happen when we have a full melt, so all tokens from the inputs are destroyed
-                # So we manually add this token to the array and set the token_data properly
-                serialized['tokens'].append(token_uid.hex())
-                output['decoded']['token_data'] = len(serialized['tokens'])
+                for out in tx.outputs:
+                    out_token_uid = tx.get_token_uid(out.get_token_index())
+                    if out_token_uid == token_uid:
+                        output['decoded']['token_data'] = out.token_data
+                        break
+                else:
+                    # This is the case when the token from the input does not appear in the outputs
+                    # This case can happen when we have a full melt, so all tokens from the inputs are destroyed
+                    # So we manually add this token to the array and set the token_data properly
+                    serialized['tokens'].append(token_uid.hex())
```

Comment:
What if the token from inputs 0 and 1 do not appear in the outputs? I feel we will duplicate the token in the `serialized['tokens']` for no reason, and it is not supposed to happen.

Why do we search for the token index in the outputs? Isn't it better to search in the `serialized['tokens']` instead?

Reply:
It makes sense! I've fixed this and refactored the code as you suggested in the other comment below in this commit: https://github.com/HathorNetwork/hathor-core/pull/181/commits/7ca07bb1f8d4729989e286d4b6d045ea5d2667c3

What do you think?

Reply:
Great! The new code can be a bit simplified (removing duplicate code) if you first check `if token_uid not in token_uid_map`, then you update the token_uid_map if necessary, and finally you do `output['decoded']['token_data'] = token_uid_map[token_uid]`.

Reply:
Refactor done